omod/src/main/java/org/openmrs/module/owa/web/controller/AddAppController.java

```diff
@@ -86,17 +87,28 @@ public String upload(@RequestParam("file") MultipartFile file, HttpServletReques
                                         uploadedFile.delete();
                                         session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
                                     } 
-                                    else {
-                                        String contextPath = request.getScheme() + "://" + request.getServerName() + ":"
-                                                + request.getServerPort() + request.getContextPath();
-                                        appManager.installApp(uploadedFile, fileName, contextPath);
-                                        message = messageSourceService.getMessage("owa.app_installed");
-                                        session.setAttribute(WebConstants.OPENMRS_MSG_ATTR, message);
+                                    else {                                    	
+                                    	message = appManager.extractMissingRequirementsMessage(uploadedFile, appManager.getStartedModules());
+                                    	if (!("".equals(message))) {
+                                    		
+                                    		//display error message for owa apps that need requirements first installed
+                                    		message = messageSourceService.getMessage("owa.missing_app_requirements") + message;
+                                    		log.warn(message);
+                                    		session.setAttribute(WebConstants.OPENMRS_MSG_ATTR, message);
+                                    	}else {
```

Comment:
For convention purposes, I thin it would be best to add the space between the else and }